File: problem1.c
#include <stdio.h>
#include <stdlib.h>

#define BASE 2

int main( int argc, char const *argv[] ) {

	FILE *pFile;
	long int *number;
	char *numberInBinary;
	int nBits, nNumeros;

	pFile = fopen( argv[1], "r" );

	fscanf( pFile, "%d%d\n", &nBits, &nNumeros );

	printf( "Running for %d bits and %d numbers\n", nBits, nNumeros );

	number = ( long int* )malloc( nNumeros * sizeof( long int ) );
	numberInBinary = ( char* )malloc( ( nBits + 1 ) * sizeof( char ) );

	for ( int i = 0; i < nNumeros; ++i ) {
		fscanf( pFile, " %[^\n]", numberInBinary );
		numberInBinary[nBits] = '\0';
		// printf( "%s\n", numberInBinary );
		number[i] = strtol( numberInBinary, NULL, BASE );
		// printf( "%d\n", number[i] );
	}



	fclose( pFile );
	return 0;
}

int degrau_2_frascos ( long int x, int n ) {
	int raiz_n = ( int )sqrt( n );
	for ( int i = 0; i < n; i += raiz_n ) {
		/* code */
		if ( i >= x ) {
			for ( int j = i - raiz_n; j < i; j++ ) {
				/* code */
				if ( j == x ) {
					return j;
				}
			}
		}
	}
}

degrau_k_frascos( long int x, int n, int k ) {
	int raiz_kesima = ( int )pow( n, 1.0f / k );
	int inicio = 0;
	int fim = n;
	int incremento = ( int ) pow( raiz_kesima, k - 1 );
	for ( int i = 0; i < k; i++ ) {
		for ( int j = inicio; j < fim; j += incremento ) {
			if ( j >= x ) {
				if ( incremento == 1 ) {
					return j;
				}
				inicio = j - incremento;
				fim = j;
				incremento = incremento / raiz_kesima;
				break;
			}
		}
	}
}

File: problem1.py
def degrau_2_frascos (x,n):
	raiz_n = pow(n,1.0/2)
	i = 0
	while i<n:		
		if i>=x :
			j = i-raiz_n
			while j<i:
				if j == x:
					return j
				j+=1
		i += int(raiz_n)
	return -1

def degrau_k_frascos(x,n,k):
	raiz_k = pow(n,1.0/k)
	inicio = 0
	fim = n
	incremento = pow(raiz_k,k-1)
	for i in k:
		j = inicio
		while j<fim:
			if j>=x:
				if incremento==1:
					return j
				inicio = j-incremento
				fim = j
				incremento = incremento / raiz_k
				break
			j+=incremento

nBits,nNumeros = input().split()
nBits = int(nBits)
nNumeros = int(nNumeros)
n = 2**nBits
for i in range(0,nNumeros,1):
	number = int(input(),2)
	resposta = degrau_2_frascos(number,n)
	if resposta != number:
		print ("Erro for ",str(number))


